linkerd/http/retry/src/lib.rs

```diff
@@ -179,21 +195,13 @@ where
 
 // === impl HttpRetry ===
 
-impl<P, N> HttpRetry<P, N> {
-    pub fn layer() -> impl layer::Layer<N, Service = Self> + Copy {
-        layer::mk(|inner| Self {
-            inner,
-            metrics: Metrics::default(),
-            _marker: PhantomData,
-        })
-    }
-}
-
-impl<P, S> Service<http::Request<BoxBody>> for HttpRetry<P, S>
+impl<P, L, X, S> Service<http::Request<BoxBody>> for HttpRetry<P, L, X, S>
 where
     P: Policy,
     P: Param<Params>,
     P: Clone + Send + Sync + std::fmt::Debug + 'static,
+    L: Clone + std::fmt::Debug + Hash + Eq + Send + Sync + prom::encoding::EncodeLabelSet + 'static,
+    X: ExtractParam<L, http::Request<BoxBody>>,
```

Comment:
We tend to rely on ExtractParam only for Targets and less so for request parameters... Is https://github.com/linkerd/linkerd2-proxy/pull/3258/commits/69872cb5a5ed06492f023a688bf09978a2ca1411 a change that can be reviewed independently of other changes in this PR?

Reply:
i've given it a shot, but it seems difficult to extricate that change into its own PR.

to restate my understanding, based on our conversation yesterday, one approach worth trying would be using a request extension rather than an `ExtractParam<P, T>`.

at risk of a brief detour, when we say "Target", we mean the `T` types provided as parameters to e.g. `Http<T>` or `Matched<T>`, at the `NewService<T>` layer. is that right? `Request<BodyBody>` is a concrete target for _this_ `Service<T>`, which had me confused, so i wanted to restate my understanding of this after reading more where we use `ExtractParam`.

i took a swing at this, here: https://github.com/linkerd/linkerd2-proxy/tree/kate/add-hostname-label-to-http-metrics.failed-extension-attempt

i found myself running into a lot of trouble with this, unfortunately. there's still something slightly incongruent stemming from the fact that we have labels that straddle the boundary between a construction time (_route and parent references_) and call-time (_request `Parts`_).

Reply:
we resolved this in-air, and confirmed the "target" terminology. :heavy_check_mark: